pkg/metrics/bpf.go

```diff
@@ -36,55 +52,70 @@ func newbpfCollector() *bpfCollector {
 }
 
 func (s *bpfCollector) Describe(ch chan<- *prometheus.Desc) {
-	ch <- s.bpfMapsMemory
-	ch <- s.bpfProgMemory
+	prometheus.DescribeByCollect(s, ch)
 }
 
 type memoryEntry struct {
 	BytesMemlock uint64 `json:"bytes_memlock"`
 }
 
-func getMemoryUsage(typ string) (uint64, error) {
+func getBPFUsage(typ string) (uint64, uint64, error) {
 	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
 	defer cancel()
 	cmd := exec.CommandContext(ctx, "bpftool", "-j", typ, "show")
 	out, err := cmd.Output()
 	if err != nil {
-		return 0, fmt.Errorf("unable to get bpftool output: %w", err)
+		return 0, 0, fmt.Errorf("unable to get bpftool output: %w", err)
 	}
 
 	var memoryEntries []memoryEntry
 	err = json.Unmarshal(out, &memoryEntries)
 	if err != nil {
-		return 0, fmt.Errorf("unable to unmarshal bpftool output: %w", err)
+		return 0, 0, fmt.Errorf("unable to unmarshal bpftool output: %w", err)
 	}
 	var totalMem uint64
 	for _, entry := range memoryEntries {
 		totalMem += entry.BytesMemlock
 	}
-	return totalMem, nil
+	return uint64(len(memoryEntries)), totalMem, nil
 }
 
 func (s *bpfCollector) Collect(ch chan<- prometheus.Metric) {
-	mapMem, err := getMemoryUsage("map")
+	// Avoid querying BPF multiple times concurrently
+	s.mutex.Lock()
```

Comment:
Not sure if this is necessary? Should this be called only per scrape interval?

Reply:
Yeah, I've added it to prevent a potential DoS attack over the metrics endpoint. IIUC this is being called on every `GET /metrics` requests, isn't it? 🤔 

Happy to remove it if you feel this is overkill!

Reply:
@ti-mo `Collect` is called on every scrape.

Reply:
Yep, singleflight is neat for this use case!

Reply:
Thanks for making me learn about this nice feat. I've replaced my former mutex implementation with it but not sure if it makes it easier in the end 🤷 let me know if there are better ways to integrate it!